File: backend/models/machine/machine.go
package machine

import (
	"time"

	"gorm.io/gorm"
)

// Machine 机器主表（Agent注册时写入静态信息）
type Machine struct {
	ID              uint           `json:"id" gorm:"primaryKey"`
	IP              string         `json:"ip" gorm:"size:45;uniqueIndex;not null"`
	Port            int            `json:"port" gorm:"not null"`
	Status          string         `json:"status" gorm:"size:20;default:online"`
	Hostname        string         `json:"hostname" gorm:"size:100"`
	Virtualization  string         `json:"virtualization" gorm:"size:50"`
	OSName          string         `json:"os_name" gorm:"size:100"`
	OSVersion       string         `json:"os_version" gorm:"size:100"`
	Kernel          string         `json:"kernel" gorm:"size:100"`
	Arch            string         `json:"arch" gorm:"size:20"`
	CPUModel        string         `json:"cpu_model" gorm:"size:200"`
	CPUCores        int32          `json:"cpu_cores"`
	MemoryKB        int64          `json:"memory_kb"`
	SwapKB          int64          `json:"swap_kb"`
	Gateway         string         `json:"gateway" gorm:"size:45"`
	FirewallStatus  string         `json:"firewall_status" gorm:"size:20"`
	FirewallEnabled string         `json:"firewall_enabled" gorm:"size:20"`
	Timezone        string         `json:"timezone" gorm:"size:50"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `json:"-" gorm:"index"`

	Disks    []MachineDisk    `json:"disks" gorm:"foreignKey:MachineID"`
	Networks []MachineNetwork `json:"networks" gorm:"foreignKey:MachineID"`
	GPUs     []MachineGPU     `json:"gpus" gorm:"foreignKey:MachineID"`
}

// MachineDisk 磁盘信息
type MachineDisk struct {
	ID        uint   `json:"id" gorm:"primaryKey"`
	MachineID uint   `json:"machine_id" gorm:"index"`
	Device    string `json:"device" gorm:"size:50"`
	Type      string `json:"type" gorm:"size:20"`
	TotalGB   int64  `json:"total_gb"`
}

// MachineNetwork 网卡信息
type MachineNetwork struct {
	ID        uint   `json:"id" gorm:"primaryKey"`
	MachineID uint   `json:"machine_id" gorm:"index"`
	Name      string `json:"name" gorm:"size:50"`
	IP        string `json:"ip" gorm:"size:45"`
	MAC       string `json:"mac" gorm:"size:20"`
	Speed     int64  `json:"speed"`
	Status    string `json:"status" gorm:"size:10"`
}

// MachineGPU GPU信息
type MachineGPU struct {
	ID            uint   `json:"id" gorm:"primaryKey"`
	MachineID     uint   `json:"machine_id" gorm:"index"`
	Name          string `json:"name" gorm:"size:200"`
	Count         int32  `json:"count"`
	DriverVersion string `json:"driver_version" gorm:"size:50"`
}

File: frontend/src/api/machine.ts
import request from '@/utils/request'

export interface DiskInfo {
  device: string
  type: string
  total_gb: number
}

export interface NetworkInfo {
  name: string
  ip: string
  mac: string
  speed: number
  status: string
}

export interface GPUInfo {
  name: string
  count: number
  driver_version: string
}

export interface MachineInfo {
  ip: string
  port: number
  status: string
  hostname: string
  virtualization: string
  uptime_seconds: number
  system_time: string
  hardware_time: string
  os_name: string
  os_version: string
  kernel: string
  arch: string
  cpu_model: string
  cpu_cores: number
  memory_kb: number
  swap_kb: number
  gateway: string
  firewall_status: string
  firewall_enabled: string
  timezone: string
  disks: DiskInfo[]
  networks: NetworkInfo[]
  gpus: GPUInfo[] | null
}

/** 获取所有机器信息（从缓存读取） */
export function getMachinesApi(): Promise<MachineInfo[]> {
  return request.get('/machines').then((res) => res.data.data)
}

/** 同步硬件信息（触发gRPC刷新） */
export function syncHardwareApi(): Promise<MachineInfo[]> {
  return request.get('/machines/sync-hardware').then((res) => res.data.data)
}

/** 删除机器记录 */
export function deleteMachineApi(ip: string, port: number): Promise<string> {
  return request.delete(`/machines/${ip}/${port}`).then((res) => res.data.data)
}

/** 远程执行命令 */
export function execOnMachineApi(ip: string, port: number, command: string): Promise<string> {
  return request.post(`/machines/${ip}/${port}/exec`, { command }).then((res) => res.data.data)
}
